typha/pkg/syncserver/snap_precalc.go

```diff
@@ -0,0 +1,281 @@
+	snappyW := snappy.NewBufferedWriter(snap.buf)
+	progressW := progressWriter{W: snappyW}
+	encoder := gob.NewEncoder(&progressW)
+	writeMsg := func(msg any) error {
+		envelope := syncproto.Envelope{
+			Message: msg,
+		}
+		err := encoder.Encode(&envelope)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	err := writeSnapshotMessages(
+		context.Background(),
+		s.logCtx.WithField("destination", "compressed in-memory cache"),
+		snap.crumb,
+		writeMsg,
+		1000, // Allow bigger messages in the snapshot.
+	)
+	if err != nil {
+		// Shouldn't happen because we're serialising to an in-memory buffer.
+		s.logCtx.WithError(err).Panic("Failed to serialise datastore snapshot.")
+	}
+
+	err = writeMsg(syncproto.MsgDecoderRestart{
+		Message:              "End of compressed snapshot.",
+		CompressionAlgorithm: syncproto.CompressionSnappy,
+	})
+	if err != nil {
+		// Shouldn't happen because we're serialising to an in-memory buffer.
+		s.logCtx.WithError(err).Panic("Failed to serialise datastore snapshot end message.")
+	}
+
+	err = snappyW.Close() // Does Flush() for us.
+	if err != nil {
+		// Shouldn't happen because we're serialising to an in-memory buffer.
+		s.logCtx.WithError(err).Panic("Failed to close datastore snapshot.")
+	}
+
+	// Closing the multi-reader buffer signals all the waiting readers.
+	err = snap.buf.Close()
+	if err != nil {
+		// Shouldn't happen because we're serialising to an in-memory buffer.
+		s.logCtx.WithError(err).Panic("Failed to close datastore snapshot.")
+	}
+
+	s.gaugeSnapBytesRaw.Set(float64(progressW.BytesWritten))
+	snapSize := snap.buf.Len()
+	s.gaugeSnapBytesComp.Set(float64(snapSize))
+
+	// Record snapshot size so that we have a good guess for next time.
+	s.setLastSnapSize(snapSize)
+}
+
+func (s *SnappySnapshotCache) setLastSnapSize(snapSize int) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	s.lastSnapSize = snapSize
+}
+
+type snapshot struct {
+	crumb *snapcache.Breadcrumb
+	buf   *multireadbuf.MultiReaderSingleWriterBuffer
+}
+
+func (s *snapshot) sendToClient(ctx context.Context, logCtx *logrus.Entry, w io.Writer, conn WriteDeadlineSetter, writeTimeout time.Duration) error {
+	var currentWriteDeadline time.Time
+	reader := s.buf.Reader()
+
+	for ctx.Err() == nil {
+		// Optimisation: only reset the deadline if we've used a significant portion of it.
+		// Setting it seems to be fairly expensive.
+		if time.Until(currentWriteDeadline) < writeTimeout {
+			newDeadline := time.Now().Add(writeTimeout * 110 / 100)
+			err := conn.SetWriteDeadline(newDeadline)
+			if err != nil {
+				return err
+			}
+			currentWriteDeadline = newDeadline
+		}
+
+		n, err := reader.WriteTo(w)
+		if err != nil {
+			if n > 0 && os.IsTimeout(err) {
+				// Managed to write _some_ bytes, loop again to reset the timeout.  If the snapshot was
+				// very large then we might have written a big chunk of it but simply not had enough time to
+				// complete.  Only give up if we see no progress at all.
```

Comment:
Would it be worth a log here in case we're hitting this a lot? 

Seems like a branch that would in normal operation be pretty rare, but would be interesting to know about in the cases we are hitting it.